XAES-256-GCM/reference.go

```diff
@@ -0,0 +1,84 @@
+package xaes256gcm
+
+import (
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/subtle"
+	"errors"
+)
+
+const (
+	KeySize   = 32
+	NonceSize = 24
+	Overhead  = 16
+)
+
+type xaes256gcm struct {
+	c  cipher.Block
+	k1 [aes.BlockSize]byte
+}
+
+func New(key []byte) (cipher.AEAD, error) {
+	if len(key) != KeySize {
+		return nil, errors.New("xaes256gcm: bad key length")
+	}
+
+	x := new(xaes256gcm)
+
+	x.c, _ = aes.NewCipher(key)
+	x.c.Encrypt(x.k1[:], x.k1[:])
+
+	var msb byte
+	for i := len(x.k1) - 1; i >= 0; i-- {
```

Comment:
This loop looks wrong to me assuming that is should implement the same shift left operation as for example used in EAX. the msb for k1[i] is used to compute the new value of k[i]. I would expect here that the msb should be used in the next iteration in the computation of k[i-1].

Reply:
Good catch! Thank you, I'll fix this. Part of why I didn't merge this yet is that I wanted to finish writing the OpenSSL implementation based on the high-level CMAC API, to find any mistakes in the low-level Go.

Reply:
I also have an independent implementation based on the spec.

Reply:
Fixed the bug, finished the OpenSSL implementation, and added test vectors. Could you check that the vectors agree with your implementation? Thank you!